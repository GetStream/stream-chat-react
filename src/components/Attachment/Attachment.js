// @ts-check

import React from 'react';
import DefaultMedia from 'react-player';
import PropTypes from 'prop-types';

import DefaultAttachmentActions from './AttachmentActions';

import DefaultAudio from './Audio';
import DefaultCard from './Card';
import DefaultFile from './FileAttachment';

import {
  ImageComponent as DefaultImage,
  Gallery as DefaultGallery,
} from '../Gallery';

export const SUPPORTED_VIDEO_FORMATS = [
  'video/mp4',
  'video/ogg',
  'video/webm',
  'video/quicktime',
];

/**
 * @typedef {import('types').ExtendedAttachment} ExtendedAttachment
 * @typedef {Required<Pick<import('types').InnerAttachmentUIComponentProps, 'Card' | 'File' | 'Gallery' |'Image' | 'Audio' | 'Media' | 'AttachmentActions'>>} DefaultProps
 * @typedef {Omit<import('types').InnerAttachmentUIComponentProps, 'Card' | 'File' | 'Image'| 'Gallery' | 'Audio' | 'Media' | 'AttachmentActions'> & DefaultProps} AttachmentProps
 */

/**
 * @param {ExtendedAttachment} a
 */
export const isGalleryAttachment = (a) => {
  return a.type === 'gallery';
};

/**
 * @param {ExtendedAttachment} a
 */
export const isImageAttachment = (a) => {
  return a.type === 'image' && !a.title_link && !a.og_scrape_url;
};

/**
 * @param {ExtendedAttachment} a
 */
export const isMediaAttachment = (a) => {
  return (
    (a.mime_type && SUPPORTED_VIDEO_FORMATS.indexOf(a.mime_type) !== -1) ||
    a.type === 'video'
  );
};

/**
 * @param {ExtendedAttachment} a
 */
export const isAudioAttachment = (a) => {
  return a.type === 'audio';
};

/**
 * @param {ExtendedAttachment} a
 */
export const isFileAttachment = (a) => {
  return (
    a.type === 'file' ||
    (a.mime_type &&
      SUPPORTED_VIDEO_FORMATS.indexOf(a.mime_type) === -1 &&
      a.type !== 'video')
  );
};

/**
 * @param {React.ReactNode} children
 * @param {ExtendedAttachment} attachment
 * @param {string} componentType
 */
export const renderAttachmentWithinContainer = (
  children,
  attachment,
  componentType,
) => {
  let extra =
    attachment && attachment.actions && attachment.actions.length
      ? 'actions'
      : '';
  if (
    componentType === 'card' &&
    !attachment.image_url &&
    !attachment.thumb_url
  ) {
    extra = 'no-image';
  }

  return (
    <div
      className={`str-chat__message-attachment str-chat__message-attachment--${componentType} str-chat__message-attachment--${attachment.type} str-chat__message-attachment--${componentType}--${extra}`}
      key={`${attachment?.id}-${attachment.type || 'none'} `}
    >
      {children}
    </div>
  );
};

/**
 * @param {AttachmentProps} props
 */
export const renderAttachmentActions = (props) => {
  const { attachment: a, AttachmentActions, actionHandler } = props;
  if (!a.actions || !a.actions.length) {
    return null;
  }

  return (
    <AttachmentActions
      {...a}
      id={a.id || ''}
      actions={a.actions || []}
      text={a.text || ''}
      key={`key-actions-${a.id}`}
      actionHandler={actionHandler}
    />
  );
};

/**
 * @param {AttachmentProps} props
 */
export const renderGallery = (props) => {
  const { attachment: a, Gallery } = props;
  return renderAttachmentWithinContainer(
    <Gallery images={a.images || []} key="gallery" />,
    a,
    'gallery',
  );
};

/**
 * @param {AttachmentProps} props
 */
export const renderImage = (props) => {
  const { attachment: a, Image } = props;
  if (a.actions && a.actions.length) {
    return renderAttachmentWithinContainer(
      <div className="str-chat__attachment" key={`key-image-${a.id}`}>
        <Image {...a} />
        {renderAttachmentActions(props)}
      </div>,
      a,
      'image',
    );
  }

  return renderAttachmentWithinContainer(
    <Image {...a} key={`key-image-${a.id}`} />,
    a,
    'image',
  );
};

/**
 * @param {AttachmentProps} props
 */
export const renderCard = (props) => {
  const { attachment: a, Card } = props;
  if (a.actions && a.actions.length) {
    return renderAttachmentWithinContainer(
      <div className="str-chat__attachment" key={`key-image-${a.id}`}>
        <Card {...a} key={`key-card-${a.id}`} />
        {renderAttachmentActions(props)}
      </div>,
      a,
      'card',
    );
  }

  return renderAttachmentWithinContainer(
    <Card {...a} key={`key-card-${a.id}`} />,
    a,
    'card',
  );
};

/**
 * @param {AttachmentProps} props
 */
export const renderFile = (props) => {
  const { attachment: a, File } = props;
  if (!a.asset_url) return null;

  return renderAttachmentWithinContainer(
    <File attachment={a} key={`key-file-${a.id}`} />,
    a,
    'file',
  );
};

/**
 * @param {AttachmentProps} props
 */
export const renderAudio = (props) => {
  const { attachment: a, Audio } = props;
  return renderAttachmentWithinContainer(
    <div className="str-chat__attachment" key={`key-video-${a.id}`}>
      <Audio og={a} />
    </div>,
    a,
    'audio',
  );
};

/**
 * @param {AttachmentProps} props
 */
export const renderMedia = (props) => {
  const { attachment: a, Media } = props;
  if (a.actions && a.actions.length) {
    return renderAttachmentWithinContainer(
      <div
        className="str-chat__attachment str-chat__attachment-media"
        key={`key-video-${a.id}`}
      >
        <div className="str-chat__player-wrapper">
          <Media
            className="react-player"
            url={a.asset_url}
            width="100%"
            height="100%"
            controls
          />
        </div>
        {renderAttachmentActions(props)}
      </div>,
      a,
      'media',
    );
  }

  return renderAttachmentWithinContainer(
    <div className="str-chat__player-wrapper" key={`key-video-${a.id}`}>
      <Media
        className="react-player"
        url={a.asset_url}
        width="100%"
        height="100%"
        controls
      />
    </div>,
    a,
    'media',
  );
};

/**
 * Attachment - The message attachment
 *
 * @example ../../docs/Attachment.md
 * @type { React.FC<import('types').WrapperAttachmentUIComponentProps> }
 */
const Attachment = ({
  attachments,
  Card = DefaultCard,
  Image = DefaultImage,
  Gallery = DefaultGallery,
  Audio = DefaultAudio,
  File = DefaultFile,
  Media = DefaultMedia,
  AttachmentActions = DefaultAttachmentActions,
  ...rest
}) => {
  const gallery = {
    type: 'gallery',
    images: attachments.filter(
      /** @param {import('types').ExtendedAttachment} a */ (a) =>
        a.type === 'image' && !(a.og_scrape_url || a.title_link),
    ),
  };
  let newAttachments;
  if (gallery.images?.length >= 2) {
    newAttachments = [
      ...attachments.filter(
        /** @param {import('types').ExtendedAttachment} a */ (a) =>
          !(a.type === 'image' && !(a.og_scrape_url || a.title_link)),
      ),
      gallery,
    ];
  } else {
    newAttachments = attachments;
  }

  const propsWithDefault = {
    Card,
    Image,
    Audio,
    File,
    Media,
    Gallery,
    AttachmentActions,
    attachments: newAttachments,
    ...rest,
  };

  return (
    <>
      {newAttachments?.map(
        /** @param {any} attachment */ (attachment) => {
          if (isGalleryAttachment(attachment)) {
            return renderGallery({ ...propsWithDefault, attachment });
          }

          if (isImageAttachment(attachment)) {
            return renderImage({ ...propsWithDefault, attachment });
          }

          if (isFileAttachment(attachment)) {
            return renderFile({ ...propsWithDefault, attachment });
          }

          if (isAudioAttachment(attachment)) {
            return renderAudio({ ...propsWithDefault, attachment });
          }

          if (isMediaAttachment(attachment)) {
            return renderMedia({ ...propsWithDefault, attachment });
          }

          return renderCard({ ...propsWithDefault, attachment });
        },
      )}
    </>
  );
};

Attachment.propTypes = {
  /**
   * The attachment to render
   * @see See [Attachment structure](https://getstream.io/chat/docs/#message_format)
   *
   *  */
  attachments: /** @type {PropTypes.Validator<ExtendedAttachment[]>} */ (PropTypes
    .array.isRequired),
  /**
   *
   * @param name {string} Name of action
   * @param value {string} Value of action
   * @param event Dom event that triggered this handler
   */
  actionHandler: PropTypes.func,
  /**
   * Custom UI component for card type attachment
   * Defaults to [Card](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/Card.js)
   */
  Card: /** @type {PropTypes.Validator<React.ComponentType<import('types').CardProps>>} */ (PropTypes.elementType),
  /**
   * Custom UI component for file type attachment
   * Defaults to [File](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/File.js)
   */
  File: /** @type {PropTypes.Validator<React.ComponentType<import('types').FileAttachmentProps>>} */ (PropTypes.elementType),
  /**
   * Custom UI component for attachment actions
   * Defaults to [AttachmentActions](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/AttachmentActions.js)
   */
  Gallery: /** @type {PropTypes.Validator<React.ComponentType<import('types').GalleryProps>>} */ (PropTypes.elementType),
  /**
   * Custom UI component for image type attachment
   * Defaults to [Image](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Gallery/Image.js)
   */
  Image: /** @type {PropTypes.Validator<React.ComponentType<import('types').ImageProps>>} */ (PropTypes.elementType),
  /**
   * Custom UI component for audio type attachment
   * Defaults to [Audio](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/Audio.js)
   */
  Audio: /** @type {PropTypes.Validator<React.ComponentType<import('types').AudioProps>>} */ (PropTypes.elementType),
  /**
   * Custom UI component for media type attachment
   * Defaults to [ReactPlayer](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/ReactPlayer.js)
   */
  Media: /** @type {PropTypes.Validator<React.ComponentType<import('react-player').ReactPlayerProps>>} */ (PropTypes.elementType),
  /**
   * Custom UI component for attachment actions
   * Defaults to [AttachmentActions](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/AttachmentActions.js)
   */
  AttachmentActions: /** @type {PropTypes.Validator<React.ComponentType<import('types').AttachmentActionsProps>>} */ (PropTypes.elementType),
};

export default Attachment;
