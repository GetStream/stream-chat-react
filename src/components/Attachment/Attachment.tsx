import React, { useMemo } from 'react';
import type { ReactPlayerProps } from 'react-player';
import type { Attachment as StreamAttachment } from 'stream-chat';

import {
  GroupedRenderedAttachment,
  isAudioAttachment,
  isFileAttachment,
  isMediaAttachment,
  isScrapedContent,
  isUploadedImage,
} from './utils';

import {
  AudioContainer,
  CardContainer,
  FileContainer,
  GalleryContainer,
  ImageContainer,
  MediaContainer,
  UnsupportedAttachmentContainer,
} from './AttachmentContainer';

import type { AttachmentActionsProps } from './AttachmentActions';
import type { AudioProps } from './Audio';
import type { CardProps } from './Card';
import type { FileAttachmentProps } from './FileAttachment';
import type { GalleryProps, ImageProps } from '../Gallery';
import type { UnsupportedAttachmentProps } from './UnsupportedAttachment';
import type { ActionHandlerReturnType } from '../Message/hooks/useActionHandler';

import type { DefaultStreamChatGenerics } from '../../types/types';

const CONTAINER_MAP = {
  audio: AudioContainer,
  card: CardContainer,
  file: FileContainer,
  media: MediaContainer,
  unsupported: UnsupportedAttachmentContainer,
} as const;

export const ATTACHMENT_GROUPS_ORDER = [
  'card',
  'gallery',
  'image',
  'media',
  'audio',
  'file',
  'unsupported',
] as const;

export type AttachmentProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  /** The message attachments to render, see [attachment structure](https://getstream.io/chat/docs/javascript/message_format/?language=javascript) **/
  attachments: StreamAttachment<StreamChatGenerics>[];
  /**	The handler function to call when an action is performed on an attachment, examples include canceling a \/giphy command or shuffling the results. */
  actionHandler?: ActionHandlerReturnType;
  /** Custom UI component for displaying attachment actions, defaults to and accepts same props as: [AttachmentActions](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/AttachmentActions.tsx) */
  AttachmentActions?: React.ComponentType<AttachmentActionsProps<StreamChatGenerics>>;
  /** Custom UI component for displaying an audio type attachment, defaults to and accepts same props as: [Audio](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/Audio.tsx) */
  Audio?: React.ComponentType<AudioProps<StreamChatGenerics>>;
  /** Custom UI component for displaying a card type attachment, defaults to and accepts same props as: [Card](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/Card.tsx) */
  Card?: React.ComponentType<CardProps>;
  /** Custom UI component for displaying a file type attachment, defaults to and accepts same props as: [File](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Attachment/FileAttachment.tsx) */
  File?: React.ComponentType<FileAttachmentProps<StreamChatGenerics>>;
  /** Custom UI component for displaying a gallery of image type attachments, defaults to and accepts same props as: [Gallery](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Gallery/Gallery.tsx) */
  Gallery?: React.ComponentType<GalleryProps<StreamChatGenerics>>;
  /** Custom UI component for displaying an image type attachment, defaults to and accepts same props as: [Image](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Gallery/Image.tsx) */
  Image?: React.ComponentType<ImageProps>;
  /** Custom UI component for displaying a media type attachment, defaults to `ReactPlayer` from 'react-player' */
  Media?: React.ComponentType<ReactPlayerProps>;
  /** Custom UI component for displaying unsupported attachment types, defaults to NullComponent */
  UnsupportedAttachment?: React.ComponentType<UnsupportedAttachmentProps>;
};

/**
 * A component used for rendering message attachments. By default, the component supports: AttachmentActions, Audio, Card, File, Gallery, Image, and Video
 */
export const Attachment = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: AttachmentProps<StreamChatGenerics>,
) => {
  const { attachments } = props;

  const groupedAttachments = useMemo(() => renderGroupedAttachments(props), [attachments]);

  return (
    <div className='str-chat__attachment-list'>
      {ATTACHMENT_GROUPS_ORDER.reduce(
        (acc, groupName) => [...acc, ...groupedAttachments[groupName]],
        [] as React.ReactNode[],
      )}
    </div>
  );
};

const renderGroupedAttachments = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>({
  attachments,
  ...rest
}: AttachmentProps<StreamChatGenerics>): GroupedRenderedAttachment => {
  const uploadedImages: StreamAttachment<StreamChatGenerics>[] = attachments.filter((attachment) =>
    isUploadedImage(attachment),
  );

  const containers = attachments
    .filter((attachment) => !isUploadedImage(attachment))
    .reduce<GroupedRenderedAttachment>(
      (typeMap, attachment) => {
        const attachmentType = getAttachmentType(attachment);

        const Container = CONTAINER_MAP[attachmentType];
        typeMap[attachmentType].push(
          <Container
            key={`${attachmentType}-${typeMap[attachmentType].length}`}
            {...rest}
            attachment={attachment}
          />,
        );

        return typeMap;
      },
      {
        audio: [],
        card: [],
        file: [],
        media: [],
        unsupported: [],
        // not used in reduce
        // eslint-disable-next-line sort-keys
        image: [],
        // eslint-disable-next-line sort-keys
        gallery: [],
      },
    );

  if (uploadedImages.length > 1) {
    containers['gallery'] = [
      <GalleryContainer
        key='gallery-container'
        {...rest}
        attachment={{
          images: uploadedImages,
          type: 'gallery',
        }}
      />,
    ];
  } else if (uploadedImages.length === 1) {
    containers['image'] = [
      <ImageContainer key='image-container' {...rest} attachment={uploadedImages[0]} />,
    ];
  }

  return containers;
};

const getAttachmentType = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  attachment: AttachmentProps<StreamChatGenerics>['attachments'][number],
): keyof typeof CONTAINER_MAP => {
  if (isScrapedContent(attachment)) {
    return 'card';
  } else if (isMediaAttachment(attachment)) {
    return 'media';
  } else if (isAudioAttachment(attachment)) {
    return 'audio';
  } else if (isFileAttachment(attachment)) {
    return 'file';
  }

  return 'unsupported';
};
