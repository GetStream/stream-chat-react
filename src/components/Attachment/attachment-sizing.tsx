import type { Attachment } from 'stream-chat';
import * as linkify from 'linkifyjs';

export const getImageAttachmentConfiguration = (attachment: Attachment, element: HTMLElement) => {
  let newUrl = undefined;

  const urlToTest = attachment.image_url || attachment.thumb_url || '';

  if (linkify.test(urlToTest, 'url')) {
    const url = new URL(urlToTest);
    const resizeDimensions = getSizingRestrictions(url, element);

    if (resizeDimensions) {
      // Apply 2x for retina displays
      resizeDimensions.height *= 2;
      resizeDimensions.width *= 2;
      addResizingParamsToUrl(resizeDimensions, url);
    }
    newUrl = url.href;
  }

  return {
    url: newUrl || '',
  };
};

export const getVideoAttachmentConfiguration = (
  attachment: Attachment,
  element: HTMLElement,
  shouldGenerateVideoThumbnail: boolean,
) => {
  let thumbUrl = undefined;
  if (
    attachment.thumb_url &&
    shouldGenerateVideoThumbnail &&
    linkify.test(attachment.thumb_url, 'url')
  ) {
    const url = new URL(attachment.thumb_url);
    const resizeDimensions = getSizingRestrictions(url, element);

    if (resizeDimensions) {
      // Apply 2x for retina displays
      resizeDimensions.height *= 2;
      resizeDimensions.width *= 2;
      addResizingParamsToUrl(resizeDimensions, url);
    }
    thumbUrl = url.href;
  }

  return {
    thumbUrl,
    url: attachment.asset_url || '',
  };
};

const getSizingRestrictions = (url: URL, htmlElement: HTMLElement) => {
  const urlParams = url.searchParams;
  const originalHeight = Number(urlParams.get('oh')) || 1;
  const originalWidth = Number(urlParams.get('ow')) || 1;
  const cssSizeRestriction = getCSSSizeRestrictions(htmlElement);
  let resizeDimensions: { height: number; width: number } | undefined;

  if ((cssSizeRestriction.maxHeight || cssSizeRestriction.height) && cssSizeRestriction.maxWidth) {
    resizeDimensions = getResizeDimensions(
      originalHeight,
      originalWidth,
      /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
      cssSizeRestriction.maxHeight || cssSizeRestriction.height!,
      cssSizeRestriction.maxWidth,
    );
  } else {
    resizeDimensions = undefined;
  }

  return resizeDimensions;
};

const getResizeDimensions = (
  originalHeight: number,
  originalWidth: number,
  maxHeight: number,
  maxWidth: number,
) => ({
  height: Math.round(Math.max(maxHeight, (maxWidth / originalWidth) * originalHeight)),
  width: Math.round(Math.max(maxHeight, (maxWidth / originalHeight) * originalWidth)),
});

const getCSSSizeRestrictions = (htmlElement: HTMLElement) => {
  const computedStylesheet = getComputedStyle(htmlElement);
  const height = getValueRepresentationOfCSSProperty(computedStylesheet.getPropertyValue('height'));
  const maxHeight = getValueRepresentationOfCSSProperty(
    computedStylesheet.getPropertyValue('max-height'),
  );
  const maxWidth = getValueRepresentationOfCSSProperty(
    computedStylesheet.getPropertyValue('max-width'),
  );

  if (!((height || maxHeight) && maxWidth)) {
    console.warn(
      `Invalid value set for height/max-height and/or max-width for HTML element, this can cause scrolling issues inside the message list, more info https://getstream.io/chat/docs/sdk/react/message-components/attachment/#image-and-video-sizing`,
    );
  }

  return { height, maxHeight, maxWidth };
};

const getValueRepresentationOfCSSProperty = (property: string) => {
  if (!property.endsWith('px')) {
    return undefined;
  }
  const number = parseFloat(property);
  return isNaN(number) ? undefined : number;
};

const addResizingParamsToUrl = (resizeDimensions: { height: number; width: number }, url: URL) => {
  url.searchParams.set('h', resizeDimensions.height.toString());
  url.searchParams.set('w', resizeDimensions.width.toString());
};
