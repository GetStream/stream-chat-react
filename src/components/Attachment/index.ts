export * from './Attachment';
export * from './AttachmentActions';
export * from './AttachmentContainer';
export * from './Audio';
export * from './Card';
export * from './UnsupportedAttachment';
export * from './FileAttachment';
export * from './utils';
