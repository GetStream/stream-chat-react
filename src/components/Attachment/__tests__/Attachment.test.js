import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { nanoid } from 'nanoid';

import {
  generateAttachmentAction,
  generateAudioAttachment,
  generateFileAttachment,
  generateGiphyAttachment,
  generateImageAttachment,
  generateScrapedAudioAttachment,
  generateScrapedDataAttachment,
  generateScrapedImageAttachment,
  generateVideoAttachment,
} from 'mock-builders';

import { Attachment } from '../Attachment';
import { SUPPORTED_VIDEO_FORMATS } from '../utils';
import { generateScrapedVideoAttachment } from '../../../mock-builders';

const Audio = (props) => <div data-testid='audio-attachment'>{props.customTestId}</div>;
const Card = (props) => <div data-testid='card-attachment'>{props.customTestId}</div>;
const Media = (props) => <div data-testid='media-attachment'>{props.customTestId}</div>;
const AttachmentActions = () => <div data-testid='attachment-actions'></div>;
const Image = (props) => <div data-testid='image-attachment'>{props.customTestId}</div>;
const File = (props) => <div data-testid='file-attachment'>{props.customTestId}</div>;
const Gallery = (props) => <div data-testid='gallery-attachment'>{props.customTestId}</div>;

const ATTACHMENTS = {
  scraped: {
    audio: generateScrapedAudioAttachment(),
    giphy: generateGiphyAttachment(),
    image: generateScrapedImageAttachment(),
    unrecognized: generateScrapedDataAttachment({ type: nanoid() }),
    video: generateScrapedVideoAttachment(),
  },
  uploaded: {
    audio: generateAudioAttachment(),
    file: generateFileAttachment(),
    gallery: Array.from({ length: 3 }, generateImageAttachment),
    image: generateImageAttachment(),
    video: generateVideoAttachment(),
  },
};

const renderComponent = (props) =>
  render(
    <Attachment
      AttachmentActions={AttachmentActions}
      Audio={Audio}
      Card={Card}
      File={File}
      Gallery={Gallery}
      Image={Image}
      Media={Media}
      {...props}
    />,
  );

describe('attachment', () => {
  describe('non-scraped content', () => {
    it('should render empty attachment list if unrecognized type', () => {
      const { container } = renderComponent({ attachments: [{}] });
      expect(container.firstChild).toBeEmptyDOMElement();
    });

    const cases = {
      audio: {
        attachments: [ATTACHMENTS.uploaded.audio],
        case: '"audio" type attachment',
        renderedComponent: 'Audio',
        testId: 'audio-attachment',
      },
      file: {
        attachments: [ATTACHMENTS.uploaded.file],
        case: '"file" type attachment',
        renderedComponent: 'File',
        testId: 'file-attachment',
      },
      gallery: {
        attachments: ATTACHMENTS.uploaded.gallery,
        case: 'message with multiple "image" attachments',
        renderedComponent: 'Gallery',
        testId: 'gallery-attachment',
      },
      image: {
        attachments: [ATTACHMENTS.uploaded.image],
        case: 'message with single image',
        renderedComponent: 'Image',
        testId: 'image-attachment',
      },
    };
    it.each`
      attachments                  | case                  | renderedComponent                  | testId
      ${cases.audio.attachments}   | ${cases.audio.case}   | ${cases.audio.renderedComponent}   | ${cases.audio.testId}
      ${cases.file.attachments}    | ${cases.file.case}    | ${cases.file.renderedComponent}    | ${cases.file.testId}
      ${cases.gallery.attachments} | ${cases.gallery.case} | ${cases.gallery.renderedComponent} | ${cases.gallery.testId}
      ${cases.image.attachments}   | ${cases.image.case}   | ${cases.image.renderedComponent}   | ${cases.image.testId}
    `('should  render $renderedComponent component for $case', async ({ attachments, testId }) => {
      renderComponent({ attachments });
      await waitFor(() => {
        expect(screen.getByTestId(testId)).toBeInTheDocument();
      });
    });

    it.each(SUPPORTED_VIDEO_FORMATS.map((f) => [f]))(
      'should render Media component for video of %s mime-type attachment',
      async (mime_type) => {
        const attachment = generateVideoAttachment({ mime_type });
        renderComponent({ attachments: [attachment] });
        await waitFor(() => {
          expect(screen.getByTestId('media-attachment')).toBeInTheDocument();
        });
      },
    );
  });

  describe('scraped content', () => {
    it('should render null, if og_scrape_url & title_link are undefined', () => {
      const attachment = generateScrapedDataAttachment({
        og_scrape_url: undefined,
        title_link: undefined,
      });
      const { container } = renderComponent({ attachments: [attachment] });
      expect(container.firstChild).toBeEmptyDOMElement();
    });

    const cases = [
      {
        attachments: [ATTACHMENTS.scraped.unrecognized],
        case: 'not recognized, but has title_link or og_scrape_url',
      },
      {
        attachments: [ATTACHMENTS.scraped.giphy],
        case: 'giphy',
      },
      {
        attachments: [ATTACHMENTS.scraped.image],
        case: 'image',
      },
      {
        attachments: [ATTACHMENTS.scraped.video],
        case: 'video',
      },
      {
        attachments: [ATTACHMENTS.scraped.audio],
        case: 'audio',
      },
    ];
    it.each`
      attachments             | case
      ${cases[0].attachments} | ${cases[0].case}
      ${cases[1].attachments} | ${cases[1].case}
      ${cases[2].attachments} | ${cases[2].case}
      ${cases[3].attachments} | ${cases[3].case}
      ${cases[4].attachments} | ${cases[4].case}
    `('should render Card if attachment type is $case', async ({ attachments }) => {
      renderComponent({ attachments });
      await waitFor(() => {
        expect(screen.getByTestId('card-attachment')).toBeInTheDocument();
      });
    });
  });

  describe('combines scraped & uploaded content', () => {
    it('should render Image and Card if one image has title_link or og_scrape_url', async () => {
      const image = generateImageAttachment();
      const card = generateScrapedImageAttachment();
      const attachments = [card, image];
      renderComponent({ attachments });
      await waitFor(() => {
        expect(screen.getByTestId('image-attachment')).toBeInTheDocument();
        expect(screen.getByTestId('card-attachment')).toBeInTheDocument();
      });
    });

    it('should render attachments in default order', async () => {
      const { container } = renderComponent({
        attachments: [
          ATTACHMENTS.uploaded.file,
          ATTACHMENTS.uploaded.audio,
          ATTACHMENTS.uploaded.video,
          ATTACHMENTS.scraped.audio,
          ATTACHMENTS.uploaded.image,
          ATTACHMENTS.uploaded.image,
        ],
      });
      await waitFor(() => {
        expect(container).toMatchSnapshot();
      });
    });

    it('should render all scraped attachments', async () => {
      const scrapedAudio = { ...ATTACHMENTS.scraped.audio, customTestId: nanoid() };
      const scrapedVideo = { ...ATTACHMENTS.scraped.video, customTestId: nanoid() };
      const scrapedImage = { ...ATTACHMENTS.scraped.image, customTestId: nanoid() };
      const { queryAllByTestId } = renderComponent({
        attachments: [
          ATTACHMENTS.uploaded.file,
          ATTACHMENTS.uploaded.audio,
          ATTACHMENTS.uploaded.video,
          scrapedAudio,
          scrapedVideo,
          scrapedImage,
          ATTACHMENTS.uploaded.image,
          ATTACHMENTS.uploaded.image,
        ],
      });
      await waitFor(() => {
        /* eslint-disable jest-dom/prefer-in-document */
        const Card = queryAllByTestId('card-attachment');

        expect(Card).toHaveLength(3);

        const idList = Array.from(Card).map((v) => v.textContent);
        expect(idList).toContain(scrapedAudio.customTestId);
        expect(idList).toContain(scrapedVideo.customTestId);
        expect(idList).toContain(scrapedImage.customTestId);
      });
    });

    it('should render all uploaded attachments', async () => {
      const { container } = renderComponent({
        attachments: [
          ATTACHMENTS.uploaded.file,
          ATTACHMENTS.uploaded.file,
          ATTACHMENTS.uploaded.audio,
          ATTACHMENTS.uploaded.audio,
          ATTACHMENTS.uploaded.video,
          ATTACHMENTS.uploaded.video,
          ATTACHMENTS.uploaded.image,
          ATTACHMENTS.uploaded.image,
        ],
      });
      await waitFor(() => {
        expect(container).toMatchSnapshot();
      });
    });
  });

  it('should render AttachmentActions component if attachment has actions', async () => {
    const action = generateAttachmentAction();
    const attachment = generateImageAttachment({
      actions: [action, action],
    });
    renderComponent({ attachments: [attachment] });
    await waitFor(() => {
      expect(screen.getByTestId('attachment-actions')).toBeInTheDocument();
    });
  });
});
