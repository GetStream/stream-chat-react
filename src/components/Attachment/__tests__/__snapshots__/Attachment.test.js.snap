// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attachment combines scraped & uploaded content should render all uploaded attachments 1`] = `
<div>
  <div
    class="str-chat__attachment-list"
  >
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--gallery"
    >
      <div
        data-testid="gallery-attachment"
      />
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--media str-chat__message-attachment--video"
    >
      <div
        class="str-chat__player-wrapper"
        data-testid="video-wrapper"
        style="--original-height: 1000000; --original-width: 1000000;"
      >
        <div
          data-testid="media-attachment"
        />
      </div>
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--media str-chat__message-attachment--video"
    >
      <div
        class="str-chat__player-wrapper"
        data-testid="video-wrapper"
        style="--original-height: 1000000; --original-width: 1000000;"
      >
        <div
          data-testid="media-attachment"
        />
      </div>
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--audio"
    >
      <div
        class="str-chat__attachment"
      >
        <div
          data-testid="audio-attachment"
        />
      </div>
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--audio"
    >
      <div
        class="str-chat__attachment"
      >
        <div
          data-testid="audio-attachment"
        />
      </div>
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--file"
    >
      <div
        data-testid="file-attachment"
      />
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--file"
    >
      <div
        data-testid="file-attachment"
      />
    </div>
  </div>
</div>
`;

exports[`attachment combines scraped & uploaded content should render attachments in default order 1`] = `
<div>
  <div
    class="str-chat__attachment-list"
  >
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--card str-chat__message-attachment--audio"
    >
      <div
        data-testid="card-attachment"
      />
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--gallery"
    >
      <div
        data-testid="gallery-attachment"
      />
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--media str-chat__message-attachment--video"
    >
      <div
        class="str-chat__player-wrapper"
        data-testid="video-wrapper"
        style="--original-height: 1000000; --original-width: 1000000;"
      >
        <div
          data-testid="media-attachment"
        />
      </div>
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--audio"
    >
      <div
        class="str-chat__attachment"
      >
        <div
          data-testid="audio-attachment"
        />
      </div>
    </div>
    <div
      class="str-chat__message-attachment str-chat__message-attachment-dynamic-size str-chat__message-attachment--file"
    >
      <div
        data-testid="file-attachment"
      />
    </div>
  </div>
</div>
`;
